File: Pwr/Sem4/JezykiIParadygmatyProgramowania/Lista5/zad3/Philosophers.java
import java.util.Random;
import java.util.concurrent.Semaphore;
import java.util.concurrent.locks.ReentrantLock;

public class Philosophers {

    static final int DEFAULT_PHILOSOPHERS = 5;
    static final int DEFAULT_MEALS = 5;

    // ===== STATYSTYKI =====
    static class Stats {
        private final int[] data;
        private final Object lock = new Object();

        Stats(int n) {
            data = new int[n];
        }

        void set(int id, int value) {
            synchronized (lock) {
                data[id] = value;
            }
        }

        int get(int id) {
            synchronized (lock) {
                return data[id];
            }
        }
    }

    // ===== FORK =====
    static class Fork {
        final ReentrantLock lock = new ReentrantLock();
    }

    // ===== WAITER =====
    static class Waiter {
        private final Semaphore seats;

        Waiter(int max) {
            seats = new Semaphore(max);
        }

        void requestSeat() throws InterruptedException {
            seats.acquire();
        }

        void leaveSeat() {
            seats.release();
        }
    }

    // ===== PHILOSOPHER =====
    static class Philosopher extends Thread {
        private final int id;
        private final Fork left;
        private final Fork right;
        private final Stats stats;
        private final Waiter waiter;
        private final Random rand;
        private final int meals;

        Philosopher(int id, Fork left, Fork right,
                     Stats stats, Waiter waiter, int meals) {
            this.id = id;
            this.left = left;
            this.right = right;
            this.stats = stats;
            this.waiter = waiter;
            this.meals = meals;
            this.rand = new Random();
        }

        void think() throws InterruptedException {
            System.out.println("Filozof " + (id + 1) + " myśli nad sensem życia.");
            Thread.sleep((rand.nextInt(3) + 1) * 100L);
        }

        void eat(int meal) throws InterruptedException {
            System.out.println("Filozof " + (id + 1)
                    + " zjada crispy chicken bacon burgera numer " + meal);
            Thread.sleep((rand.nextInt(3) + 1) * 100L);
        }

        @Override
        public void run() {
            int failed = 0;

            for (int meal = 1; meal <= meals; meal++) {
                try {
                    think();

                    System.out.println("Filozof " + (id + 1) + " jest głodny i próbuje zjeść.");
                    waiter.requestSeat();

                    boolean leftTaken = left.lock.tryLock();
                    if (!leftTaken) {
                        failed++;
                        System.out.println("Filozof " + (id + 1) + " nie podniósł lewego widelca.");
                        waiter.leaveSeat();
                        continue;
                    }

                    boolean rightTaken = right.lock.tryLock();
                    if (!rightTaken) {
                        failed++;
                        System.out.println("Filozof " + (id + 1) + " nie podniósł prawego widelca.");
                        left.lock.unlock();
                        waiter.leaveSeat();
                        continue;
                    }

                    eat(meal);

                    left.lock.unlock();
                    right.lock.unlock();
                    waiter.leaveSeat();

                } catch (InterruptedException e) {
                    Thread.currentThread().interrupt();
                }
            }

            stats.set(id, failed);
        }
    }

    // ===== MAIN =====
    public static void main(String[] args) throws InterruptedException {

        int numPhilosophers = DEFAULT_PHILOSOPHERS;
        int meals = DEFAULT_MEALS;

        if (args.length >= 2) {
            numPhilosophers = Integer.parseInt(args[0]);
            meals = Integer.parseInt(args[1]);
        }

        Fork[] forks = new Fork[numPhilosophers];
        for (int i = 0; i < numPhilosophers; i++) {
            forks[i] = new Fork();
        }

        Stats stats = new Stats(numPhilosophers);
        Waiter waiter = new Waiter(numPhilosophers - 1);

        Philosopher[] philosophers = new Philosopher[numPhilosophers];

        for (int i = 0; i < numPhilosophers; i++) {
            philosophers[i] = new Philosopher(
                    i,
                    forks[i],
                    forks[(i + 1) % numPhilosophers],
                    stats,
                    waiter,
                    meals
            );
            philosophers[i].start();
        }

        for (Philosopher p : philosophers) {
            p.join();
        }

        System.out.println("\nStatystyki");
        for (int i = 0; i < numPhilosophers; i++) {
            System.out.println("Filozof " + (i + 1)
                    + " tyle razy obszedł się smakiem = " + stats.get(i));
        }
    }
}

File: Pwr/Sem4/JezykiIParadygmatyProgramowania/Lista5/zad2/philosophers.go
package main

import (
	"fmt"
	"math/rand"
	"os"
	"strconv"
	"sync"
	"time"
)

const DefaultNumPhilosophers = 5
const DefaultMealsPerPhilosopher = 5

type Stats struct {
	mu   sync.Mutex
	data []int
}

func NewStats(n int) *Stats {
	return &Stats{
		data: make([]int, n),
	}
}

func (s *Stats) Set(id int, value int) {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.data[id] = value
}

func (s *Stats) Get(id int) int {
	s.mu.Lock()
	defer s.mu.Unlock()
	return s.data[id]
}

type Fork struct {
	mu sync.Mutex
}

type Waiter struct {
	seat chan struct{}
}

func NewWaiter(max int) *Waiter {
	return &Waiter{
		seat: make(chan struct{}, max),
	}
}

func (w *Waiter) RequestSeat() {
	w.seat <- struct{}{}
}

func (w *Waiter) LeaveSeat() {
	<-w.seat
}

type Philosopher struct {
	id     int
	left   *Fork
	right  *Fork
	stats  *Stats
	waiter *Waiter
	rng    *rand.Rand
}

func (p *Philosopher) think() {
	fmt.Printf("Filozof %d myśli nad sensem życia.\n", p.id+1)
	time.Sleep(time.Duration(p.rng.Intn(3)+1) * 100 * time.Millisecond)
}

func (p *Philosopher) eat(meal int) {
	fmt.Printf("Filozof %d zjada burgera numer %d\n",
		p.id+1, meal)
	time.Sleep(time.Duration(p.rng.Intn(3)+1) * 100 * time.Millisecond)
}

func (p *Philosopher) run(meals int, wg *sync.WaitGroup) {
	defer wg.Done()

	failed := 0

	for meal := 1; meal <= meals; meal++ {
		p.think()

		fmt.Printf("Filozof %d jest głodny i próbuje zjeść.\n", p.id+1)
		p.waiter.RequestSeat()

		leftCh := make(chan bool, 1)
		rightCh := make(chan bool, 1)

		go func() {
			locked := p.left.mu.TryLock()
			leftCh <- locked
		}()

		select {
		case ok := <-leftCh:
			if !ok {
				failed++
				fmt.Printf("Filozof %d nie podniósł lewego widelca.\n", p.id+1)
				p.waiter.LeaveSeat()
				continue
			}
		case <-time.After(100 * time.Millisecond):
			failed++
			fmt.Printf("Filozof %d za wolny (lewy widelec).\n", p.id+1)
			p.waiter.LeaveSeat()
			continue
		}

		go func() {
			locked := p.right.mu.TryLock()
			rightCh <- locked
		}()

		select {
		case ok := <-rightCh:
			if !ok {
				failed++
				fmt.Printf("Filozof %d nie podniósł prawego widelca.\n", p.id+1)
				p.left.mu.Unlock()
				p.waiter.LeaveSeat()
				continue
			}
		case <-time.After(100 * time.Millisecond):
			failed++
			fmt.Printf("Filozof %d za wolny (prawy widelec).\n", p.id+1)
			p.left.mu.Unlock()
			p.waiter.LeaveSeat()
			continue
		}

		p.eat(meal)

		p.left.mu.Unlock()
		p.right.mu.Unlock()
		p.waiter.LeaveSeat()
	}

	p.stats.Set(p.id, failed)
}

func main() {
	numPhilosophers := DefaultNumPhilosophers
	meals := DefaultMealsPerPhilosopher

	if len(os.Args) >= 3 {
		numPhilosophers, _ = strconv.Atoi(os.Args[1])
		meals, _ = strconv.Atoi(os.Args[2])
	}

	rng := rand.New(rand.NewSource(time.Now().UnixNano()))

	forks := make([]*Fork, numPhilosophers)
	for i := range forks {
		forks[i] = &Fork{}
	}

	stats := NewStats(numPhilosophers)
	waiter := NewWaiter(numPhilosophers - 1)

	var wg sync.WaitGroup

	for i := 0; i < numPhilosophers; i++ {
		p := &Philosopher{
			id:     i,
			left:   forks[i],
			right:  forks[(i+1)%numPhilosophers],
			stats:  stats,
			waiter: waiter,
			rng:    rand.New(rand.NewSource(rng.Int63())),
		}

		wg.Add(1)
		go p.run(meals, &wg)
	}

	wg.Wait()

	fmt.Println("\nStatystyki")
	for i := 0; i < numPhilosophers; i++ {
		fmt.Printf("Filozof %d tyle razy obszedł się smakiem = %d\n",
			i+1, stats.Get(i))
	}
}